src/combinatorics/simple_graph/basic.lean

```diff
@@ -94,39 +97,107 @@ by { classical, exact fintype.of_injective simple_graph.adj simple_graph.ext }
 
 @[simp]
 lemma simple_graph.from_rel_adj {V : Type u} (r : V → V → Prop) (v w : V) :
-  (simple_graph.from_rel r).adj v w ↔ v ≠ w ∧ (r v w ∨ r w v) :=
-iff.rfl
+  (simple_graph.from_rel r).adj v w ↔ v ≠ w ∧ (r v w ∨ r w v) := iff.rfl
 
-/--
-The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent.
--/
-def complete_graph (V : Type u) : simple_graph V :=
-{ adj := ne }
+/-- The complete graph on a type `V` is the simple graph with all pairs of distinct vertices
+adjacent. In `mathlib`, this is usually referred to as `⊤`. -/
+def complete_graph (V : Type u) : simple_graph V := { adj := ne }
+
+/-- The graph with no edges on a given vertex type `V`. `mathlib` prefers the notation `⊥`. -/
+def empty_graph (V : Type u) : simple_graph V := { adj := λ i j, false }
+
+namespace simple_graph
+
+variables {V : Type u} {W : Type v} {X : Type w} (G : simple_graph V) (G' : simple_graph W)
+
+@[simp] lemma irrefl {v : V} : ¬G.adj v v := G.loopless v
+
+lemma adj_comm (u v : V) : G.adj u v ↔ G.adj v u := ⟨λ x, G.sym x, λ x, G.sym x⟩
+
+@[symm] lemma adj_symm {u v : V} (h : G.adj u v) : G.adj v u := G.sym h
+
+lemma ne_of_adj {a b : V} (hab : G.adj a b) : a ≠ b :=
+by { rintro rfl, exact G.irrefl hab }
+
+section order
+
+/-- The relation that one `simple_graph` is a subgraph of another. -/
+def is_subgraph (x y : simple_graph V) : Prop := ∀ ⦃v w : V⦄, x.adj v w → y.adj v w
+
+/-- The union of two `simple_graph`s. -/
+protected def union (x y : simple_graph V) : simple_graph V :=
+{ adj := x.adj ⊔ y.adj,
+  sym := λ v w h, by rwa [sup_apply, sup_apply, x.adj_comm, y.adj_comm] }
+
+/-- The intersection of two `simple_graph`s. -/
+protected def inter (x y : simple_graph V) : simple_graph V :=
+{ adj := x.adj ⊓ y.adj,
+  sym := λ v w h, by rwa [inf_apply, inf_apply, x.adj_comm, y.adj_comm] }
+
+/-- We define `compl G` to be the `simple_graph V` such that no two adjacent vertices in `G`
+are adjacent in the complement, and every nonadjacent pair of vertices is adjacent
+(still ensuring that vertices are not adjacent to themselves.) -/
+protected def compl (G : simple_graph V) : simple_graph V :=
+{ adj := λ v w, v ≠ w ∧ ¬G.adj v w,
+  sym := λ v w ⟨hne, _⟩, ⟨hne.symm, by rwa adj_comm⟩,
+  loopless := λ v ⟨hne, _⟩, false.elim (hne rfl) }
 
-instance (V : Type u) : inhabited (simple_graph V) :=
-⟨complete_graph V⟩
+/-- The difference of two `simple_graph`s, which is the simple graph whose edges
+are all of those from the first graph that aren't in the second graph. -/
+protected def sdiff (x y : simple_graph V) : simple_graph V :=
+{ adj := x.adj \ y.adj,
+  sym := λ v w h, by change x.adj w v ∧ ¬ y.adj w v; rwa [x.adj_comm, y.adj_comm]  }
```

Comment:
I would lean towards defining these directly as the instances:
```suggestion
/-- We define `Gᶜ` to be the `simple_graph V` such that no two adjacent vertices in `G`
are adjacent in the complement, and every nonadjacent pair of vertices is adjacent
(still ensuring that vertices are not adjacent to themselves.) -/
instance : has_compl (simple_graph V) :=
⟨{ adj := λ v w, v ≠ w ∧ ¬G.adj v w,
   sym := λ v w ⟨hne, _⟩, ⟨hne.symm, by rwa adj_comm⟩,
   loopless := λ v ⟨hne, _⟩, false.elim (hne rfl) }⟩
  
/-- The difference of two `simple_graph`s, which is the simple graph whose edges
are all of those from the first graph that aren't in the second graph. -/
instance : has_sdiff (simple_graph V) :=
⟨{ adj := x.adj \ y.adj,
   sym := λ v w h, by change x.adj w v ∧ ¬ y.adj w v; rwa [x.adj_comm, y.adj_comm]  }⟩
```
etc. The corresponding `foo_adj` lemmas can go immediately after each instance.

Reply:
shame `simps` makes something like `simple_graph.has_sup_sup_adj:
        > ∀ {V : Type u} (x y : simple_graph V) (ᾰ ᾰ_1 : V), (x ⊔ y).adj ᾰ ᾰ_1 = (x.adj ⊔ y.adj) ᾰ ᾰ_1` here, so I have to do it manually

Reply:
`@[simps {simp_rhs := tt}]` does a better job probably, but still uses a weird lemma name.

Reply:
it makes `(x.adj ᾰ ∪ y.adj ᾰ) ᾰ_1` actually, which doesn't seem like the right simp-normal form

Reply:
I didn't know we also had `pi.has_union`

Reply:
That's a little alarming - it must be swapping between `set X` and `X -> Prop` somewhere...

Does this fail if you put it in that file?
```
example {α} [has_sup α] (x y : α → α → Prop) (m n : α) : (x ⊔ y) m n :=
begin
  simp?,
  guard_target_strict x m n ∨ y m n,
  sorry
end
```

Reply:
yep, the `simp?` output is `simp only [set.sup_eq_union, sup_apply]`

Reply:
Ouch, that shouldn't happen :(